File: magix/magix-java-client/src/main/java/ru/mipt/npm/magix/client/MagixClient.java
package ru.mipt.npm.magix.client;

import kotlinx.serialization.json.JsonElement;
import ru.mipt.npm.magix.api.MagixMessage;

import java.io.IOException;
import java.util.concurrent.Flow;

/**
 * See https://github.com/waltz-controls/rfc/tree/master/2
 *
 * @param <T>
 */
public interface MagixClient<T> {
    void broadcast(MagixMessage msg) throws IOException;

    Flow.Publisher<MagixMessage> subscribe();

    /**
     * Create a magix endpoint client using RSocket with raw tcp connection
     * @param host host name of magix server event loop
     * @param port port of magix server event loop
     * @return the client
     */
    static MagixClient<JsonElement> rSocketTcp(String host, int port) {
        return ControlsMagixClient.Companion.rSocketTcp(host, port);
    }

    /**
     *
     * @param host host name of magix server event loop
     * @param port port of magix server event loop
     * @param path context path for WS connection
     */
    static MagixClient<JsonElement> rSocketWs(String host, int port, String path) {
        return ControlsMagixClient.Companion.rSocketWs(host, port, path);
    }
}

File: controls-magix-client/src/commonMain/kotlin/ru/mipt/npm/controls/client/magixClient.kt
package space.kscience.dataforge.control.client

import kotlinx.coroutines.Job
import kotlinx.coroutines.flow.catch
import kotlinx.coroutines.flow.launchIn
import kotlinx.coroutines.flow.onEach
import kotlinx.coroutines.launch
import ru.mipt.npm.magix.api.MagixEndpoint
import space.kscience.dataforge.context.error
import space.kscience.dataforge.context.logger
import space.kscience.dataforge.control.controllers.DeviceManager
import space.kscience.dataforge.control.controllers.respondMessage
import space.kscience.dataforge.control.messages.DeviceMessage
import space.kscience.dataforge.magix.api.MagixMessage


public const val DATAFORGE_MAGIX_FORMAT: String = "dataforge"

private fun generateId(request: MagixMessage<DeviceMessage>): String = if (request.id != null) {
    "${request.id}.response"
} else {
    "df[${request.payload.hashCode()}"
}

/**
 * Communicate with server in [Magix format](https://github.com/waltz-controls/rfc/tree/master/1)
 */
public fun DeviceManager.launchMagixClient(
    endpoint: MagixEndpoint<DeviceMessage>,
    endpointID: String = DATAFORGE_MAGIX_FORMAT,
): Job = context.launch {
    endpoint.subscribe().onEach { request ->
        //TODO analyze action

        val responsePayload = respondMessage(request.payload)
        val response = MagixMessage(
            format = DATAFORGE_MAGIX_FORMAT,
            id = generateId(request),
            parentId = request.id,
            origin = endpointID,
            payload = responsePayload
        )
        endpoint.broadcast(response)
    }.catch { error ->
        logger.error(error) { "Error while responding to message" }
    }.launchIn(this)

    controller.messageOutput().onEach { payload ->
        MagixMessage(
            format = DATAFORGE_MAGIX_FORMAT,
            id = "df[${payload.hashCode()}]",
            origin = endpointID,
            payload = payload
        )
    }.catch { error ->
        logger.error(error) { "Error while sending a message" }
    }.launchIn(this)
}


